alfi/solver.py

```diff
@@ -37,6 +37,41 @@ def apply(self, pc, x, y):
     def applyTranspose(self, pc, x, y):
         raise NotImplementedError("Sorry!")
 
+class CGMassInv(PCBase):
+
+    def initialize(self, pc):
+        _, P = pc.getOperators()
+        prefix = pc.getOptionsPrefix()
+        appctx = self.get_appctx(pc)
+        V = dmhooks.get_function_space(pc.getDM())
+        # get function spaces
+        u = TrialFunction(V)
+        v = TestFunction(V)
+        mass = assemble(inner(u, v)*dx)
+
+        Mksp = PETSc.KSP().create(comm=pc.comm)
+        Mksp.incrementTabLevel(1, parent=pc)
+        Mksp.setOptionsPrefix(prefix + "Mp_")
+        Mksp.setOperators(mass.petscmat)
+        Mksp.setUp()
```

Comment:
Shouldn't we call `setFromOptions` _before_ `setUp`? It's possible your CG mass matrix options aren't taking effect. Or if they are, you're setting up the default preconditioner (GMRES/ILU?) and then throwing it away.

Reply:
This is copied pretty much straight from [pcd.py](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/preconditioners/pcd.py) in firedrake. But you're right, the [petsc documentation](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/KSP/KSPSetFromOptions.html) also seems to suggest that `KSPSetFromOptions` should be called first. @wence- , can you chime in here?

Reply:
Ok, we should do what petsc does/recommends